Retrofit rx adapter

DSMISHA opened the pull request:
The search option is not implemented. It will be completed bit later.

DarkSpirit111 committed:
TVshows screen was implemented

DarkSpirit111 committed:
TVshows modifications

DarkSpirit111 committed:
Movie details screen: part of ui was implemented

DarkSpirit111 committed:
various modifications after pull request feedback

DarkSpirit111 committed:
various modifications after pull request feedback

DarkSpirit111 committed:
kestore gitignore

DarkSpirit111 committed:
API: movies, tvShows requests were implemented

DarkSpirit111 committed:
API: small modifications

DarkSpirit111 committed:
API: "watching now" section was implemented

DarkSpirit111 committed:
API: details was implemented
some optimisations were added
- search is not implemented

AndroidStudentClub commented on app/build.gradle:
```diff
@@ -77,4 +79,18 @@ dependencies {
     implementation "com.jakewharton.timber:timber:$rootProject.timberVersion"
+
+    // RxJava2 Dependencies
+    implementation 'io.reactivex.rxjava2:rxjava:2.2.10'
```
Рекомендуется выносить версии библиотек. Пример :$rootProject.recyclerViewVersion"
Если интересно, то вот более продвинутая статья https://androidschool.ru/2019/10/09/improve-build-gradle/

AndroidStudentClub commented on app/build.gradle:
```diff
@@ -77,4 +79,18 @@ dependencies {
+    implementation "com.squareup.retrofit2:converter-gson:2.5.0"
+
+    implementation 'com.squareup.okhttp3:okhttp:3.14.9'
+    implementation 'com.squareup.okhttp3:logging-interceptor:3.6.0'
```
Рекомендуется выносить версии библиотек

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/MovieApiClient.kt:
```diff
@@ -1,5 +1,39 @@
+
+    val apiClient: MovieApiInterface by lazy{
+        val gson = GsonBuilder()
+            .setLenient()
```
С одной стороны прикольная фича, с другой стороны есть такой принцип Fail fast - то есть если что то работает не так - то лучше узнать об этом как можно раньше. C этим параметром - можно долго потом искать причину какой-нибудь ошибки. Так что использовать стоит с осторожностью

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/MovieApiInterface.kt:
```diff
@@ -1,6 +1,56 @@
 
+    @GET("movie/popular")
+    fun getPopularMovies(
+        @Query("api_key") apiKey: String,
```
API key можно в дефолтные аргументы - чтобы не прокидывать каждый раз

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/MovieApiInterface.kt:
```diff
@@ -1,6 +1,56 @@
+    @GET("movie/popular")
+    fun getPopularMovies(
+        @Query("api_key") apiKey: String,
+        @Query("language") lang: String,
```
Язык можно впринципе тоже в дефолтные прокинуть

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/MovieApiInterface.kt:
```diff
@@ -1,6 +1,56 @@
+        @Query("api_key") apiKey: String,
+        @Query("language") lang: String,
+        @Query("page") page: Int
+    ): Observable<MovieWrapper?>?
```
Обычно в сетевых запросах рекомендуется использовать Single

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/MovieApiInterface.kt:
```diff
@@ -1,6 +1,56 @@
+        @Query("api_key") apiKey: String,
+        @Query("language") lang: String,
+        @Query("page") page: Int
+    ): Observable<MovieWrapper?>?
```
Обычно в сетевых запросах рекомендуется использовать Single

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/MovieApiInterface.kt:
```diff
@@ -1,6 +1,56 @@
+        @Query("api_key") apiKey: String,
+        @Query("language") lang: String,
+        @Query("page") page: Int
+    ): Observable<TvShowWrapper?>?
```
Обычно в сетевых запросах рекомендуется использовать Single

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/RestApi.kt:
```diff
@@ -0,0 +1,55 @@
+import java.io.IOException
+
+interface RestApiInterface{
+    fun getPopularMovies(): Observable<MovieWrapper?>?
```
Обычно в сетевых запросах рекомендуется использовать Single

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/RestApi.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    override fun getNewMovies(): Observable<MovieWrapper?>?  {
+        val lang = "en-US"
```
Думаю можно это все в дефолтные аргументы закинуть, чтобы не повторять

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/RestApi.kt:
```diff
@@ -0,0 +1,55 @@
+object RestApi : RestApiInterface{
+
+    override fun getPopularMovies(): Observable<MovieWrapper?>?  {
+        val lang = "en-US"
```
Думаю можно это все в дефолтные аргументы закинуть, чтобы не повторять

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/network/RestApi.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    override fun getTvShows(): Observable<TvShowWrapper?>?  {
+        val lang = "en-US"
```
Думаю можно это все в дефолтные аргументы закинуть, чтобы не повторять

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/SearchBar.kt:
```diff
@@ -46,6 +49,23 @@ class SearchBar @JvmOverloads constructor(
+            }
+
+            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
+                textChangeListener?.onTextChange(p0.toString())
```
Есть ряд удобных kotlin extension, чтобы добавить их надо добавить implementation 'androidx.core:core-ktx:1.0.0'

Потом можно просто сделать вот так EditText.doAfterTextChanged{ } и не писать пустые колбэки

AndroidStudentClub replied:
Надо ещё через Observable.create() сделать - чтобы реализовать поиск в реактивном стиле. Тогда Listener не нужен будет

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -33,25 +37,79 @@ class FeedFragment : Fragment() {
+            ?.subscribeWith(object : WatchingObserver(){})
+
+        val dis3 = MovieFinderApp.instance?.getRestApi()?.getNewMovies()?.subscribeOn(Schedulers.io())
+            ?.observeOn(AndroidSchedulers.mainThread())
```
Вот это 
```
subscribeOn(Schedulers.io())
            ?.observeOn(AndroidSchedulers.mainThread())
```

Можно в экстеншн чтобы не дублировать. Или через compose()

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -33,25 +37,79 @@ class FeedFragment : Fragment() {
+
+    private fun loadDataAndInitMovieSections(){
+        val dis = MovieFinderApp.instance?.getRestApi()?.getPopularMovies()?.subscribeOn(Schedulers.io())
+            ?.observeOn(AndroidSchedulers.mainThread())
```
Можно в экстеншн чтобы не дублировать subscirebOn/ObserveOn

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -33,25 +37,79 @@ class FeedFragment : Fragment() {
+        dis3?.let { compoDisposable.add(dis3) }
+    }
+
+    private open inner class PopularObserver : SimpleSubscriber<MovieWrapper>(){
```
По сути, все эти 3 класса  делают одно и тоже. Отличие только в заголовке категории. Поэтому можно оставить один  и передавать туда название категории

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -33,25 +37,79 @@ class FeedFragment : Fragment() {
+
+    private open inner class WatchingObserver : SimpleSubscriber<MovieWrapper>(){
+        override fun onNext(it: MovieWrapper) {
+            if (it.result != null) {
```
По котлиновски лучше isNotNull()

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -60,27 +118,20 @@ class FeedFragment : Fragment() {
-        adapter.apply { addAll(newMoviesList) }
 
+    private fun initSearchBar() {
+        search_toolbar.search_edit_text.afterTextChanged {
```
когда будет через Observable - то можно будет тут подписаться и добавить фильтрацию через операторы RxJava

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/MovieItem.kt:
```diff
@@ -4,26 +4,45 @@ import com.squareup.picasso.Picasso
-            .load("https://m.media-amazon.com/images/M/MV5BYTk3MDljOWQtNGI2My00OTEzLTlhYjQtOTQ4ODM2MzUwY2IwXkEyXkFqcGdeQXVyNTIzOTk5ODM@._V1_.jpg")
-            .into(viewHolder.image_preview)
+    private fun setRating(viewHolder: GroupieViewHolder){
+        content.voteAverage?.let { viewHolder.movieRating.rating = it/2 }
```
Деление рейтинга лучше сделать прямо в модели через property get() - таким образом инкапсулируем логику для этой модели прямо в модели. И если где то еще понадобится, то достаточно будет использовать это свойство модели

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/MovieItem.kt:
```diff
@@ -4,26 +4,45 @@ import com.squareup.picasso.Picasso
+    private fun loadImage(viewHolder: GroupieViewHolder){
+        if(!content.poster.isNullOrEmpty()) {
+            Picasso.get()
+                .load(BuildConfig.BASE_POSTER_URL + content.poster)
```
Вот это
`BuildConfig.BASE_POSTER_URL + content.poster`
тоже лучше реализовать прямо в модели через свойство

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/feed/MovieItem.kt:
```diff
@@ -4,26 +4,45 @@ import com.squareup.picasso.Picasso
     }
+
+    private fun loadImage(viewHolder: GroupieViewHolder){
+        if(!content.poster.isNullOrEmpty()) {
```
Можно через let - так будет больше соответствовать котлин стилю

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/CharacterItem.kt:
```diff
@@ -0,0 +1,36 @@
+    private fun loadImage(viewHolder: GroupieViewHolder){
+        if(!content.image.isNullOrEmpty()) {
+            Picasso.get()
+                .load(BuildConfig.BASE_POSTER_URL + content.image)
```
Вот это
`BuildConfig.BASE_POSTER_URL + content.image`
тоже лучше реализовать прямо в модели через свойство и здесь передавать уже просто content.image - внутри которого будет конкатенация

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+    }
+
+    private fun loadDetailsData(){
+        if(movieId == null) return
```
isNull()

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+        if(movieId == null) return
+
+        val dis = MovieFinderApp.instance?.getRestApi()?.getDetails(movieId!!)?.subscribeOn(Schedulers.io())
+            ?.observeOn(AndroidSchedulers.mainThread())
```
```
.subscribeOn(Schedulers.io())
            ?.observeOn(AndroidSchedulers.mainThread())
```

В экстеншн

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+        tv_studio.text = getProductionString()
+        tv_genre.text = getGenresString()
+        tv_year.text = movie?.releaseDate?.substring(0, 4)
+        movie?.voteAverage?.let { movie_rating_details.rating = it/2 }
```
Вот тоже повторяется. Если бы вынесли в свойство модели - то уже не надо повторять было бы. 
Так что лучше вынести в модель логику it/2

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+    private fun loadImage(){
+        if(!movie?.poster.isNullOrEmpty()) {
+            Picasso.get()
+                .load(BuildConfig.BASE_POSTER_URL + movie?.poster)
```
Конкатенация должна быть в модели 
`BuildConfig.BASE_POSTER_URL + movie?.poster`

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+
+    private fun getProductionString(): String{
+        var sProd = ""
+        movie?.productionCompanies?.forEach {
```
Этот метод тоже бы в модель - так как он относится к модели

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+
+    private fun getGenresString(): String{
+        var sGenres = ""
+        movie?.genres?.forEach {
```
Тоже было бы неплохо прямо в модель

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+            CharacterItem(it)
+        }?.toList()
+
+        if(charactersList != null) {
```
isNotNull()

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -31,15 +62,117 @@ class MovieDetailsFragment : Fragment() {
+
+
+    private fun onWatchClicked(){
+        btn_watch.setOnClickListener {
```
Удобней здесь было бы использовать Checkbox c кастомным селектором

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/tvshows/Tvshowitem.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private fun setRating(viewHolder: GroupieViewHolder){
+        content.voteAverage?.let { viewHolder.movie_rating.rating = it/2 }
```
3 раз повторяется) В модель нужно

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/tvshows/Tvshowitem.kt:
```diff
@@ -0,0 +1,70 @@
+    private fun loadImage(viewHolder: GroupieViewHolder){
+        if(!content.poster.isNullOrEmpty()) {
+            Picasso.get()
+                .load(BuildConfig.BASE_POSTER_URL + content.poster)
```
В модель

AndroidStudentClub commented on app/src/main/java/ru/mikhailskiy/intensiv/ui/tvshows/TvshowsFragment.kt:
```diff
@@ -0,0 +1,116 @@
+        tvshows_recycler_view.layoutManager = LinearLayoutManager(context)
+        if(adapter.itemCount == 0){
+            disposable = MovieFinderApp.instance?.getRestApi()?.getTvShows()?.subscribeOn(Schedulers.io())
+                ?.observeOn(AndroidSchedulers.mainThread())
```
```
subscribeOn(Schedulers.io())
                ?.observeOn(AndroidSchedulers.mainThread())
```
В экстеншн

AndroidStudentClub commented on app/src/main/res/layout/movie_details_fragment.xml:
```diff
@@ -1,13 +1,184 @@
+
+        <FrameLayout
+            android:id="@+id/btn_watch"
+            android:layout_width="150dp"
```
размеры/цвета/стили лучше выносить в ресурсы

AndroidStudentClub commented on app/src/main/res/values/colors.xml:
```diff
@@ -8,6 +8,7 @@
 
     <color name="blueDark">#2C3848</color>
 
+    <color name="semi_white">#88FFFFFF</color>]
```
лишняя скобка ]

AndroidStudentClub reviewed (changes requested):
Проверил. Желательно вынести в экстеншны повторяющийся код. Вынести логику в модели

DarkSpirit111 committed:
pull request feedback
- dependencies versions were moved to dependencies.gradle

DarkSpirit111 committed:
pull request feedback
- GsonBuilder serLenient was commented

DarkSpirit111 committed:
pull request feedback
- "language" and "api key" quarry params for API requests were moved to the api interface

DarkSpirit111 committed:
pull request feedback
- the repeating logic "averageRating/2" was moved from adapters to models as a delegate

DarkSpirit111 committed:
pull request feedback
- height param from xml layout was exported to the dimen
